Fix dummy disconnecting on hot reload

KebsCS opened the pull request:
Fixes #8903

I'm not sure if the 1 frame jump when swapping back to the player can be fixed, as the client needs to be swapped to a dummy for it to be initialized

https://github.com/user-attachments/assets/a2a4fe4a-bf16-4936-8c7b-3cbd3e017fd2



- [x] Tested the change ingame
- [x] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test (especially base/) or added coverage to integration test
- [ ] Considered possible null pointers and out of bounds array indexing
- [ ] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

Robyt3 reviewed (changes requested):
This implementation is too brittle and broad for me. It adds code paths that intentionally trigger the bug #2804. The issue #8903 was about using the `reload` command, but this PR also makes dummy reconnect whenever the next map has the same name as the previous (like using `change_map`).

Using the map name from the map details message is brittle as well. Not every server sends map details messages, only the map change message is guaranteed. This could lead to strange dummy connecting behavior when connecting to different servers (of which some send map details messages and some do not). Especially because the respective variables are not currently being reset in `DisconnectWithReason`.

Some ideas for alternative solutions:

1. We add a client command `dummy_reconnect` which makes the client automatically reconnect the dummy when connecting to the next server. Then a bind for `dummy_reconnect; rcon reload` could be used in the client to reload and reconnect the dummy. This would need a discussion whether we want to support automatic dummy reconnecting in general.
2. Make this feature more specific to the editor. Before executing the `hot_reload` command after saving the map in the editor, set a flag in the client that the dummy should be reconnected the next time. This flag would be identical to the `dummy_reconnect` command, but it wouldn't be exposed as a command to players and only affect map reloading triggered by saving the map in the editor.
3. Add a new server message that tells the client that the map is being reloaded, which causes the client to reconnect the dummy when next ingame again.

CC @furo321

furo321 commented:
I was more thinking that the dummy's connection would stay connected instead of reconnecting. This would also guarantee that they're given the same client id as before if there are multiple dummies connected to the server. Not sure how hard this would be to implement though.

> 3. Add a new server message that tells the client that the map is being reloaded, which causes the client to reconnect the dummy when next ingame again.

This seems like the best option if you ask me.

KebsCS commented:
@Robyt3 Done, implemented 3.
It will only reconnect the dummy on the 'reload' and 'hot_reload' commands. I also added a force to the DummyConnect, because when the map was reloaded shortly after the dummy was connected, it would fail, only showing a silent error in the console. I don't think it breaks anything

KebsCS commented:
@Robyt3 Done, now dummy won't disconnect on map reload and works the same as the player

Robyt3 commented on src/engine/client/client.h:
```diff
@@ -182,6 +182,8 @@ class CClient : public IClient, public CDemoPlayer::IListener
 	bool m_DummyConnecting = false;
 	bool m_DummyConnected = false;
 	float m_LastDummyConnectTime = 0.0f;
+	bool m_DummyReconnectOnReload = false;
+	bool m_DummySwapOnReconnect = false;
```
These variables should be reset also in the `DisconnectWithReason` function, so the dummy reconnecting behavior does not activate if the client disconnects after getting the map reload but before the map is reloaded.

Robyt3 reviewed (approved):
Thanks, looks good to me!
